Create rule S1874: Deprecated code should not be used in Kubernetes (#3593)
SonarSource/rspec: rules/S1874/kubernetes/rule.adoc

Reviewer: Please add developer instructions.

Reviewer 2: is not there for other languages in this rule. Should it?

Reviewer: Yes, but they apparently forgot it. But just focus on our rule.

== Why is this an issue?

include::../description.adoc[]

> === Code examples

Reviewer: Either we should add `How to fix it` section or we should change the code examples to contain compliant and noncompliant code example in one box, like in the other languages of this rule.
See https://github.com/SonarSource/rspec/blob/master/docs/description.adoc -> Why is this an issue? and How to fix it section.

Reviewer: I'm fine with both, but would slightly prefer the second version.

Reviewer 2: I will prefer to align it with other languages as well. I will check and update accordingly

==== Noncompliant code example

> Consider the following example that uses `.Capabilities.KubeVersion.GitVersion` which is deprecated. The recommendation is to use `.Capabilities.KubeVersion.Version` instead.

Reviewer: ```suggestion
Use `.Capabilities.KubeVersion.Version` instead of the deprecated call to `.Capabilities.KubeVersion.GitVersion`
```

[source,text,diff-id=1,diff-type=noncompliant]
----
apiVersion: v1
kind: ConfigMap
metadata:
  name: {{ .Release.Name }}-configmap
data:
  myvalue: "Hello World"
  kubeVersion: {{ .Capabilities.KubeVersion.GitVersion }} # Noncompliant
----

==== Compliant solution

[source,text,diff-id=1,diff-type=compliant]
----
apiVersion: v1
kind: ConfigMap
metadata:
  name: {{ .Release.Name }}-configmap
data:
  myvalue: "Hello World"
  kubeVersion: {{ .Capabilities.KubeVersion.Version }} # Compliant
----

include::../see.adoc[]

> * Go Documentation - https://pkg.go.dev/helm.sh/helm/v3/pkg/chartutil#KubeVersion.GitVersion[GitVersion deprecated]

Reviewer: Nice find!

```suggestion
* Go Documentation - https://pkg.go.dev/helm.sh/helm/v3/pkg/chartutil#KubeVersion.GitVersion[Deprecated KubeVersion.GitVersion]
```

== Why is this an issue?

include::../description.adoc[]

> Example: Use `.Capabilities.KubeVersion.Version` instead of the deprecated call to `.Capabilities.KubeVersion.GitVersion`

Reviewer: Why example? We only have one deprecated call, so it's ok to be more tailored to it.
As a side task, maybe you can look if there are other deprecated methods in K8s?

[source,text]
----
apiVersion: v1
kind: ConfigMap
metadata:
  name: {{ .Release.Name }}-configmap
data:
  myvalue: "Hello World"
  kubeVersion: {{ .Capabilities.KubeVersion.GitVersion }} # Noncompliant

apiVersion: v1
kind: ConfigMap
metadata:
  name: {{ .Release.Name }}-configmap
data:
  myvalue: "Hello World"

>   kubeVersion: {{ .Capabilities.KubeVersion.Version }}

Reviewer: This is not really how I imagined both non compliant and compliant in one box. This doesn't make it more readable.

In the other languages, both calls are used in one class or example. Try to combine the to.
Maybe add a `kubeVersionDeprecated` property next to `kubeVersion`.

----

include::../see.adoc[]

* Go Documentation - https://pkg.go.dev/helm.sh/helm/v3/pkg/chartutil#KubeVersion.GitVersion[Deprecated KubeVersion.GitVersion]

ifdef::env-github,rspecator-view[]

'''
== Implementation Specification
(visible only on this page)

=== Message

> Show the message added along with commented tag for the deprecated code. For the above example - Deprecated: use KubeVersion.Version.

Reviewer: I don't understand this, as this doesn't specify a clear message I should use here.
Would you be able to know exactly what to use as a message if you were the developer?

Reviewer: Remove this deprecated use of `KubeVersion.GitVersion`, use `KubeVersion.Version` instead.
